src/main/java/org/broadinstitute/hellbender/utils/bigquery/StorageAPIAvroReader.java

```diff
@@ -0,0 +1,161 @@
+                    .setTableReference(tableReference)
+                    .setReadOptions(tableReadOptions)
+                    .setRequestedStreams(1)
+                    .setFormat(Storage.DataFormat.AVRO);
+
+            final Storage.ReadSession session = client.createReadSession(builder.build());
+            Preconditions.checkState(session.getStreamsCount() > 0);
+
+            this.schema = new org.apache.avro.Schema.Parser().parse(session.getAvroSchema().getSchema());
+
+            this.datumReader = new GenericDatumReader<>(
+                    new org.apache.avro.Schema.Parser().parse(session.getAvroSchema().getSchema()));
+
+            // Use the first stream to perform reading.
+            Storage.StreamPosition readPosition = Storage.StreamPosition.newBuilder()
+                    .setStream(session.getStreams(0))
+                    .build();
+
+            Storage.ReadRowsRequest readRowsRequest = Storage.ReadRowsRequest.newBuilder()
+                    .setReadPosition(readPosition)
+                    .build();
+
+            this.serverStream = client.readRowsCallable().call(readRowsRequest).iterator();
+
+            loadNextRow();
+        } catch ( IOException e ) {
+            throw new GATKException("I/O Error", e);
+        }
+    }
+
+    private void loadNextRow() {
+        try {
+            if ( decoder != null && ! decoder.isEnd() ) {
+                nextRow = datumReader.read(null, decoder);
+            } else {
+                fetchNextAvroRows();
+
+                if ( decoder != null && ! decoder.isEnd() ) {
+                    nextRow = datumReader.read(null, decoder);
+                } else {
+                    nextRow = null; // end of traversal
+                }
+            }
+        } catch ( IOException e ) {
+            throw new GATKException("I/O error", e);
+        }
+    }
+
+    private void fetchNextAvroRows() {
+        if ( serverStream.hasNext() ) {
+            currentAvroRows = serverStream.next().getAvroRows();
+            decoder = DecoderFactory.get()
+                    .binaryDecoder(currentAvroRows.getSerializedBinaryRows().toByteArray(), decoder);
+        } else {
+            currentAvroRows = null;
+            decoder = null;
+        }
+    }
+
+    @Override
+    public org.apache.avro.Schema getSchema() {
+        return schema;
+    }
+
+    @Override
+    public Iterator<GenericRecord> iterator() {
+        return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextRow != null;
+    }
+
+    @Override
+    public GenericRecord next() {
+        if ( ! hasNext() ) {
+            throw new NoSuchElementException("next() called when ! hasNext()");
+        }
+
+        final GenericRecord recordToReturn = nextRow;
+        loadNextRow();
+        return recordToReturn;
+    }
+
+    @Override
+    public void close() {
+        client.shutdownNow();
+        /* TODO: do we need to wait for termination here?
```

Comment:
Do we know the answer to this?

Reply:
I don't think we know the answer to this at this point, so could we leave the TODO for now?